Support for Kubernetes-native Spark clusters (`k8s://`)

I would like to be able to submit jobs with the kubernetes-native Spark scheduler:
https://github.com/apache-spark-on-k8s/spark

Using their spark distribution/release, it is so easy to spin up a kubernetes cluster (e.g. in Google GKE) and immediately start running jobs, observing, logs, etc.

For example, submitting a job this way:
```
kubectl proxy &
spark-submit \
  --deploy-mode=cluster \
  --kubernetes-namespace=spark \
  --master=k8s://http://127.0.0.1:8001 \
  ...
 ```

I believe when BigDL initializes (`init_engine`), it fails on validating the `--master` format with `k8s://` in it, not sure how many other things would be involved in adding support for a new scheduler in BigDL, but it would be great to be able to use native kubernetes.

I was able to get a bit further by making the modification (based on 0.2 branch) here:
https://github.com/intel-analytics/BigDL/compare/branch-0.2...paulreimer:add-spark-on-k8s-scheduler-support

At this point I can submit BigDL jobs (although these are my first, I'm not sure what to expect), but I do see drivers and executors started and the logs look promising up until this point:

```
spark-submit --conf spark.executor.instances=2 --conf spark.app.name=lenet5 --properties-file=/tmp/spark-bigdl.conf --conf spark.cores.max=4 --conf spark.executor.cores=2 --jars gs://spark-resource-staging/bigdl-0.2.0-jar-with-dependencies.jar gs://spark-resource-staging/lenet5.py
...
2017-10-19 16:25:09 INFO  DAGScheduler:54 - Final stage: ResultStage 7 (reduce at DistriOptimizer.scala:260)
2017-10-19 16:25:09 INFO  DAGScheduler:54 - Parents of final stage: List(ShuffleMapStage 6)
2017-10-19 16:25:09 INFO  DAGScheduler:54 - Missing parents: List()
2017-10-19 16:25:09 INFO  DAGScheduler:54 - Submitting ResultStage 7 (ZippedPartitionsRDD2[31] at zipPartitions at DistriOptimizer.scala:156), which has no missing parents
2017-10-19 16:25:09 INFO  MemoryStore:54 - Block broadcast_7 stored as values in memory (estimated size 10.3 KB, free 408.6 MB)
2017-10-19 16:25:09 INFO  MemoryStore:54 - Block broadcast_7_piece0 stored as bytes in memory (estimated size 5.2 KB, free 408.6 MB)
2017-10-19 16:25:09 INFO  BlockManagerInfo:54 - Added broadcast_7_piece0 in memory on lenet5-1508430265731-driver-svc.spark.svc.cluster.local:7079 (size: 5.2 KB, free: 408.8 MB)
2017-10-19 16:25:09 INFO  SparkContext:54 - Created broadcast 7 from broadcast at DAGScheduler.scala:1006
2017-10-19 16:25:09 INFO  DAGScheduler:54 - Submitting 2 missing tasks from ResultStage 7 (ZippedPartitionsRDD2[31] at zipPartitions at DistriOptimizer.scala:156) (first 15 tasks are for partitions Vector
(0, 1))
2017-10-19 16:25:09 INFO  KubernetesTaskSchedulerImpl:54 - Adding task set 7.0 with 2 tasks
2017-10-19 16:25:09 INFO  KubernetesTaskSetManager:54 - Starting task 1.0 in stage 7.0 (TID 10, 10.56.1.13, executor 1, partition 1, PROCESS_LOCAL, 5271 bytes)
2017-10-19 16:25:09 INFO  KubernetesTaskSetManager:54 - Starting task 0.0 in stage 7.0 (TID 11, 10.56.0.19, executor 2, partition 0, PROCESS_LOCAL, 5271 bytes)
2017-10-19 16:25:09 INFO  BlockManagerInfo:54 - Added broadcast_7_piece0 in memory on 10.56.0.19:46689 (size: 5.2 KB, free: 314.0 MB)
2017-10-19 16:25:09 INFO  BlockManagerInfo:54 - Added broadcast_7_piece0 in memory on 10.56.1.13:46511 (size: 5.2 KB, free: 314.0 MB)
2017-10-19 16:25:09 INFO  BlockManagerInfo:54 - Added rdd_27_0 in memory on 10.56.0.19:46689 (size: 117.2 KB, free: 313.9 MB)
2017-10-19 16:25:09 INFO  BlockManagerInfo:54 - Added rdd_27_1 in memory on 10.56.1.13:46511 (size: 117.2 KB, free: 313.9 MB)
2017-10-19 16:25:09 WARN  KubernetesTaskSetManager:66 - Lost task 0.0 in stage 7.0 (TID 11, 10.56.0.19, executor 2): java.util.concurrent.ExecutionException: java.lang.ClassCastException: org.apache.spark
.storage.ByteBufferBlockData cannot be cast to org.apache.spark.util.io.ChunkedByteBuffer
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.util.concurrent.FutureTask.get(FutureTask.java:192)
        at com.intel.analytics.bigdl.parameters.FutureResult$$anonfun$waitResult$1.apply(AllReduceParameter.scala:305)
        at com.intel.analytics.bigdl.parameters.FutureResult$$anonfun$waitResult$1.apply(AllReduceParameter.scala:305)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        at scala.collection.Iterator$class.foreach(Iterator.scala:893)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
        at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
        at scala.collection.AbstractTraversable.map(Traversable.scala:104)
        at com.intel.analytics.bigdl.parameters.FutureResult.waitResult(AllReduceParameter.scala:305)
        at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$6.apply(DistriOptimizer.scala:184)
        at com.intel.analytics.bigdl.optim.DistriOptimizer$$anonfun$6.apply(DistriOptimizer.scala:157)
        at org.apache.spark.rdd.ZippedPartitionsRDD2.compute(ZippedPartitionsRDD.scala:89)
        at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
        at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
        at org.apache.spark.scheduler.Task.run(Task.scala:108)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: org.apache.spark.storage.ByteBufferBlockData cannot be cast to org.apache.spark.util.io.ChunkedByteBuffer
        at com.intel.analytics.bigdl.parameters.AllReduceParameter$$anonfun$1$$anon$2.call(AllReduceParameter.scala:188)
        at com.intel.analytics.bigdl.parameters.AllReduceParameter$$anonfun$1$$anon$2.call(AllReduceParameter.scala:184)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        ... 3 more
```

Does this look like it is at least close to being able to complete the `lenet5` job? I'm not sure where to look next, or if this is something that would be already working on the `master` / 0.3.0 line (which is not yet available in `conda`, which is why I based my changes of 0.2.0).

@yiheng Yup, I am using Spark 2.2. I did try rebasing my small change onto latest `master`, and I was able to get a bit further:
https://github.com/intel-analytics/BigDL/compare/master...paulreimer:add-spark-on-k8s-scheduler-support

```
creating: createSpatialConvolution
creating: createTanh
creating: createSpatialMaxPooling
creating: createTanh
creating: createSpatialConvolution
creating: createSpatialMaxPooling
creating: createReshape
creating: createLinear
creating: createTanh
creating: createLinear
creating: createLogSoftMax
creating: createClassNLLCriterion
creating: createDefault
creating: createSGD
creating: createMaxEpoch
creating: createOptimizer
disableCheckSingleton is deprecated. Please use bigdl.check.singleton instead
creating: createEveryEpoch
creating: createTop1Accuracy
creating: createEveryEpoch
2017-10-20 16:12:15 INFO  DistriOptimizer$:756 - caching training rdd ...
2017-10-20 16:12:18 ERROR KubernetesTaskSetManager:70 - Task 2 in stage 0.0 failed 4 times; aborting job
Traceback (most recent call last):
  File "/var/spark-data/spark-files/lenet5.py", line 105, in <module>
    trained_model = optimizer.optimize()
  File "/opt/spark/python/lib/bigdl-0.3.0-SNAPSHOT-python-api.zip/bigdl/optim/optimizer.py", line 598, in optimize
  File "/opt/spark/python/lib/bigdl-0.3.0-SNAPSHOT-python-api.zip/bigdl/util/common.py", line 477, in callJavaFunc
  File "/opt/spark/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py", line 1133, in __call__
  File "/opt/spark/python/lib/py4j-0.10.4-src.zip/py4j/protocol.py", line 319, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o252.optimize.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 2 in stage 0.0 failed 4 times, most recent failure: Lost task 2.3 in stage 0.0 (TID 12, 10.56.0.11, executor 2): java.lang.ClassCastException: cannot assign instance of scala.collection.immutable.List$SerializationProxy to field org.apache.spark.rdd.RDD.org$apache$spark$rdd$RDD$$dependencies_ of type scala.collection.Seq in instance of org.apache.spark.rdd.MapPartitionsRDD
        at java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(ObjectStreamClass.java:2133)
        at java.io.ObjectStreamClass.setObjFieldValues(ObjectStreamClass.java:1305)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2251)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2169)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2245)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2169)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:422)
        at scala.collection.immutable.List$SerializationProxy.readObject(List.scala:479)
        at sun.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1058)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2136)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2245)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2169)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2245)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2169)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:422)
        at org.apache.spark.serializer.JavaDeserializationStream.readObject(JavaSerializer.scala:75)
        at org.apache.spark.serializer.JavaSerializerInstance.deserialize(JavaSerializer.scala:114)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:85)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
        at org.apache.spark.scheduler.Task.run(Task.scala:108)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
        at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1499)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1487)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1486)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1486)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
        at scala.Option.foreach(Option.scala:257)
        at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:814)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1714)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1669)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1658)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
        at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:630)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2022)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2043)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2062)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2087)
        at org.apache.spark.rdd.RDD.count(RDD.scala:1158)
        at com.intel.analytics.bigdl.dataset.DistributedDataSet$$anon$5.cache(DataSet.scala:188)
        at com.intel.analytics.bigdl.optim.DistriOptimizer.prepareInput(DistriOptimizer.scala:757)
        at com.intel.analytics.bigdl.optim.DistriOptimizer.optimize(DistriOptimizer.scala:777)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
        at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
        at py4j.Gateway.invoke(Gateway.java:280)
        at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
        at py4j.commands.CallCommand.execute(CallCommand.java:79)
        at py4j.GatewayConnection.run(GatewayConnection.java:214)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: cannot assign instance of scala.collection.immutable.List$SerializationProxy to field org.apache.spark.rdd.RDD.org$apache$spark$rdd$RDD$$dependencies_ of type scala.collection.Seq in instance of org.apache.spark.rdd.MapPartitionsRDD
        at java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(ObjectStreamClass.java:2133)
        at java.io.ObjectStreamClass.setObjFieldValues(ObjectStreamClass.java:1305)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2251)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2169)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2245)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2169)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:422)
        at scala.collection.immutable.List$SerializationProxy.readObject(List.scala:479)
        at sun.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1058)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2136)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2245)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2169)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2245)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2169)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2027)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1535)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:422)
        at org.apache.spark.serializer.JavaDeserializationStream.readObject(JavaSerializer.scala:75)
        at org.apache.spark.serializer.JavaSerializerInstance.deserialize(JavaSerializer.scala:114)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:85)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
        at org.apache.spark.scheduler.Task.run(Task.scala:108)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        ... 1 more

Exception in thread "main" org.apache.spark.SparkUserAppException: User application exited with 1
        at org.apache.spark.deploy.PythonRunner$.main(PythonRunner.scala:97)
        at org.apache.spark.deploy.PythonRunner.main(PythonRunner.scala)
```

--- a/docs/docs/ProgrammingGuide/run-on-k8s.md
+++ b/docs/docs/ProgrammingGuide/run-on-k8s.md
@@ -28,7 +28,7 @@ The images are as follows:
 |PySpark Driver Image|intelanalytics/spark-driver-py:v2.2.0-kubernetes-0.5.0|
 |PySpark Executor Image|intelanalytics/spark-executor-py:v2.2.0-kubernetes-0.5.0|
 
-You may also build your own customized images.
+You may also build your own customized images. see instructions [here](https://github.com/intel-analytics/BigDL/tree/master/docker/BigDL).
 
 ---
 ## **Run BigDL examples**

--- a/docs/docs/ProgrammingGuide/run-on-k8s.md
+++ b/docs/docs/ProgrammingGuide/run-on-k8s.md
@@ -1,0 +1,85 @@
+---
+# **Run BigDL on Kubernetes**
+
+The [Apache Spark on Kubernetes](https://apache-spark-on-k8s.github.io/userdocs/index.html) project enables
+native support for submitting Spark application to a kubernetes cluster. As a deep learning library for Apache
+Spark, BigDL can also run on Kubernetes by leveraging Spark on Kubernetes.
+
+---
+## **Prerequisites**
+
+1. You need to have a running Kubernetes cluster that support Spark on Kubernetes. See [here](https://apache-spark-on-k8s.github.io/userdocs/running-on-kubernetes.html#prerequisites)
+
+2. You need to spin up the **resource staging server** for dependency management. See [Dependency Management](https://apache-spark-on-k8s.github.io/userdocs/running-on-kubernetes.html#dependency-management) (This is optional if all your application dependencies are
+packaged into your own custom docker image or resides in remote locations like HDFS. See [Dependency Management Without The Resource Staging Server](https://apache-spark-on-k8s.github.io/userdocs/running-on-kubernetes.html#dependency-management-without-the-resource-staging-server))
+
+---
+## **Docker images**
+
+BigDL already published pre-built docker images that can be deployed into containers with pods.
+
+The images are as follows:
+
+|Component|Image|
+|---|---|
+|Spark Driver Image|intelanalytics/spark-driver:v2.2.0-kubernetes-0.5.0|
+|Spark Executor Image|intelanalytics/spark-executor:v2.2.0-kubernetes-0.5.0|
+|Spark Initialization Image|intelanalytics/spark-init:v2.2.0-kubernetes-0.5.0|
+|PySpark Driver Image|intelanalytics/spark-driver-py:v2.2.0-kubernetes-0.5.0|
+|PySpark Executor Image|intelanalytics/spark-executor-py:v2.2.0-kubernetes-0.5.0|
+
+You may also build your own customized images.
+
+---
+## **Run BigDL examples**
+
+Run BigDL on Kubernetes is quite easy once you meet the prerequisites above. For example,
+to run the BigDL scala Lenet example:
+
+```shell
+SPARK_HOME=...
+BIGDL_HOME=...
+$SPARK_HOME/bin/spark-submit \
+  --deploy-mode cluster \
+  --class com.intel.analytics.bigdl.models.lenet.Train \
+  --master k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port> \
+  --kubernetes-namespace default \
+  --conf spark.executor.instances=4 \
+  --conf spark.app.name=bigdl-lenet \
+  --conf spark.executor.cores=1 \
+  --conf spark.cores.max=4 \
+  --conf spark.kubernetes.driver.docker.image=intelanalytics/spark-driver:v2.2.0-kubernetes-0.5.0-ubuntu-14.04 \
+  --conf spark.kubernetes.executor.docker.image=intelanalytics/spark-executor:v2.2.0-kubernetes-0.5.0-ubuntu-14.04 \
+  --conf spark.kubernetes.initcontainer.docker.image=intelanalytics/spark-init:v2.2.0-kubernetes-0.5.0-ubuntu-14.04 \
+  --conf spark.kubernetes.resourceStagingServer.uri=http://<address-of-any-cluster-node>:31000 \
+  $BIGDL_HOME/lib/bigdl-0.4.0-SNAPSHOT-jar-with-dependencies.jar \
+-f hdfs://master:9000/mnist \
+-b 128 \
+-e 2 \
+--checkpoint /tmp
+```
+
+To run python lenet example:
+
+```shell
+SPARK_HOME=...
+BIGDL_HOME=...
+$SPARK_HOME/bin/spark-submit \
+  --deploy-mode cluster \
+  --master k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port> \
+  --kubernetes-namespace default \
+  --jars $BIGDL_HOME/lib/bigdl-0.4.0-SNAPSHOT-jar-with-dependencies.jar \
+  --py-files $BIGDL_HOME/lib/bigdl-0.4.0-SNAPSHOT-python-api.zip \
+  --conf spark.executor.instances=4 \
+  --conf spark.app.name=bigdl-1 \
+  --conf spark.executor.cores=1 \
+  --conf spark.cores.max=4 \
+  --conf spark.kubernetes.driver.docker.image=intelanalytics/spark-driver-py:v2.2.0-kubernetes-0.5.0-ubuntu-14.04 \
+  --conf spark.kubernetes.executor.docker.image=intelanalytics/spark-executor-py:v2.2.0-kubernetes-0.5.0-ubuntu-14.04 \
+  --conf spark.kubernetes.initcontainer.docker.image=intelanalytics/spark-init:v2.2.0-kubernetes-0.5.0-ubuntu-14.04 \
+  --conf spark.kubernetes.resourceStagingServer.uri=http://<address-of-any-cluster-node>:31000 \
+  bigdl/models/lenet/lenet5.py \
+  --action train \
+  --dataPath /tmp/mnist \
+  -n 2
+```

--- a/pyspark/bigdl/util/common.py
+++ b/pyspark/bigdl/util/common.py
@@ -464,7 +464,7 @@ def create_spark_conf():
         extend_spark_driver_cp(sparkConf, get_bigdl_classpath())
 
     # add content in PYSPARK_FILES in spark.submit.pyFiles
-    # need this for k8s
+    # This is a workaround for current Spark on k8s
     python_lib = os.environ.get('PYSPARK_FILES', None)
     if python_lib:
         existing_py_files = sparkConf.get("spark.submit.pyFiles")

--- a/pyspark/bigdl/util/common.py
+++ b/pyspark/bigdl/util/common.py
@@ -462,6 +462,17 @@ def create_spark_conf():
     sparkConf.setAll(bigdl_conf.items())
     if not is_spark_below_2_2():
         extend_spark_driver_cp(sparkConf, get_bigdl_classpath())
+
+    # add content in PYSPARK_FILES in spark.submit.pyFiles
+    # need this for k8s
+    python_lib = os.environ.get('PYSPARK_FILES', None)
+    if python_lib:
+        existing_py_files = sparkConf.get("spark.submit.pyFiles")
+        if existing_py_files:
+            sparkConf.set(key="spark.submit.pyFiles", value="%s,%s" % (python_lib, existing_py_files))
+        else:
+            sparkConf.set(key="spark.submit.pyFiles", value=python_lib)
+
     return sparkConf
 
 

--- a/spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala
+++ b/spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala
@@ -426,9 +426,9 @@ object Engine {
       val coreString = conf.get("spark.executor.cores", null)
       val maxString = conf.get("spark.cores.max", null)
       require(coreString != null, "Engine.init: Can't find executor core number" +
-        ", do you submit with --executor-cores option")
+        ", do you submit with --conf spark.executor.cores option")
       require(maxString != null, "Engine.init: Can't find total core number" +
-        ". Do you submit with --total-executor-cores")
+        ". Do you submit with --conf spark.cores.max option")
       val core = coreString.toInt
       val nodeNum = dynamicAllocationExecutor(conf).getOrElse {
         val total = maxString.toInt

--- a/spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala
+++ b/spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala
@@ -421,6 +421,23 @@ object Engine {
         total / core
       }
       Some(nodeNum, core)
+    } else if (master.toLowerCase.startsWith("k8s")) {
+      // Spark-on-kubernetes mode
+      val coreString = conf.get("spark.executor.cores", null)
+      val maxString = conf.get("spark.cores.max", null)
+      require(coreString != null, "Engine.init: Can't find executor core number" +
+        ", do you submit with --executor-cores option")
+      require(maxString != null, "Engine.init: Can't find total core number" +
+        ". Do you submit with --total-executor-cores")
+      val core = coreString.toInt
+      val nodeNum = dynamicAllocationExecutor(conf).getOrElse {
+        val total = maxString.toInt
+        require(total >= core && total % core == 0, s"Engine.init: total core " +
+          s"number($total) can't be divided " +
+          s"by single core number($core) provided to spark-submit")
+        total / core
+      }
+      Some(nodeNum, core)
     } else {
       throw new IllegalArgumentException(s"Engine.init: Unsupported master format $master")
     }

